server/src/WebsocketServer.hs

```diff
@@ -40,22 +41,12 @@ newUUID = randomIO
 broadcast :: [Text] -> Value -> ServerState -> IO ()
 broadcast =
   let
-    send :: WS.Connection -> Value -> IO ()
-    send conn value =
-      WS.sendTextData conn (Aeson.encode value)
-      `catch`
-      sendFailed
-
-    sendFailed :: SomeException -> IO ()
-    sendFailed exc
-      -- Rethrow async exceptions, they are meant for inter-thread communication
-      -- (e.g. ThreadKilled) and we don't expect them at this point.
-      | Just asyncExc <- fromException exc = throwIO (asyncExc :: SomeAsyncException)
-      -- We want to catch all other errors in order to prevent them from
-      -- bubbling up and disrupting the broadcasts to other clients.
-      | otherwise = pure ()
+    writeToSubQueue :: TBQueue Value -> Value -> IO ()
+    writeToSubQueue queue val = atomically $ do
+      isFull <- isFullTBQueue queue
+      unless isFull $ writeTBQueue queue val
```

Comment:
For a production ready version, it would be interesting to count how many updates we lose due to individual queues being full (so just having a counter metric here for the `isFull` case).
That allows us to measure whether the chosen queue length is right (just a few lost events occasionally) or too small (many events lost).

Reply:
I think it would also be good to log a message every time we discard a message, due to the queue being full.

Reply:
> What extra information would a log give us in this case that we wouldn't get from a counter metric?

I would log the actual message that is being discarded (which you wouldn't get from the counter). I think this is useful for debugging, since you would get a direct log message and otherwise you would need to know that IP discards messages sometimes and would need to check the counter metric.

> I would expect such a log to be rather spammy.

I think it shouldn't be too bad, since we should just close the TCP connection to any customer that is timing out (@DiegoDiverio is that what happens now btw? Is the timeout value configurable?). So you would only get some log spam until the connection is closed.

Reply:
> I would log the actual message that is being discarded (which you wouldn't get from the counter). I think this is useful for debugging, since you would get a direct log message and otherwise you would need to know that IP discards messages sometimes and would need to check the counter metric.

I don't think that would help with debugging, as a client being too slow to receive data (or just a connection that was silently dropped) is not a problem we can reasonably take action on.

----

Adding a metric does have value though, because it allows us to actually observe that things are working correctly. (Logs would to of course, but a metric is both easier to observe and has lower overhead). That's why I would also disagree with

> Though I would advise against getting this PR too large: we should solve a problem at a time to keep the PR readable

The change in this PR is pretty big in terms of how icepeak operates, and leaves us with no way of observing that it is working correctly, and not losing half the updates or something. (We would probably eventually notice if it affects what our users see - but that's too late)

A simple counter metric would be a low overhead way of achieving this: as long as the metric is zero (or rising very slowly) it means that the updates are still working as intended.

If you want to keep the PR small, you can create a second PR to be based on this one, then merge them together before deploying.

Reply:
Ok, makes sense. Let's go with the counter metric.

Reply:
Yeah, that's exactly the dev pattern I was looking for. I won't deploy before having merged all the developments we talked about here.